src/coins.h

```diff
@@ -127,9 +152,59 @@ struct CCoinsCacheEntry
         FRESH = (1 << 1),
     };
 
-    CCoinsCacheEntry() : flags(0) {}
-    explicit CCoinsCacheEntry(Coin&& coin_) : coin(std::move(coin_)), flags(0) {}
-    CCoinsCacheEntry(Coin&& coin_, unsigned char flag) : coin(std::move(coin_)), flags(flag) {}
+    CCoinsCacheEntry() noexcept = default;
+    explicit CCoinsCacheEntry(Coin&& coin_) noexcept : coin(std::move(coin_)) {}
+    ~CCoinsCacheEntry()
+    {
+        ClearFlags();
+    }
+
+    //! Adding a flag also requires a self reference to the pair that contains
+    //! this entry in the CCoinsCache map and a reference to the sentinel of the
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
+    {
+        Assume(&self.second == this);
+        if (!m_flags && flags) {
+            m_prev = sentinel.second.m_prev;
+            m_next = &sentinel;
+            sentinel.second.m_prev = &self;
+            m_prev->second.m_next = &self;
```

Comment:
there's a hidden setter and updater here, which is applicable in SelfRef as well, since you need to the commits anyway, please consider extracting the repetition:
```patch
diff --git a/src/coins.h b/src/coins.h
--- a/src/coins.h	(revision 28f0dd2d60d0f9a8d8a5aa0ed4a5eb1bd4f605bf)
+++ b/src/coins.h	(date 1722697977399)
@@ -128,6 +128,16 @@
     CoinsCachePair* m_next{nullptr};
     uint8_t m_flags{0};
 
+    void SetPrevNext(CoinsCachePair* prev, CoinsCachePair* next) noexcept {
+        m_prev = prev;
+        m_next = next;
+    }
+
+    void UpdateLink(CoinsCachePair* prev, CoinsCachePair* next) noexcept {
+        m_prev->second.m_next = next;
+        m_next->second.m_prev = prev;
+    }
+
 public:
     Coin coin; // The actual cached data.
 
@@ -166,18 +176,16 @@
     {
         Assume(&self.second == this);
         if (!m_flags && flags) {
-            m_prev = sentinel.second.m_prev;
-            m_next = &sentinel;
-            sentinel.second.m_prev = &self;
-            m_prev->second.m_next = &self;
+            SetPrevNext(sentinel.second.m_prev, &sentinel);
+            UpdateLink(&self, &self);
         }
         m_flags |= flags;
     }
     inline void ClearFlags() noexcept
     {
-        if (!m_flags) return;
-        m_next->second.m_prev = m_prev;
-        m_prev->second.m_next = m_next;
+        if (m_flags) {
+            UpdateLink(m_prev, m_next);
+        }
         m_flags = 0;
     }
     inline uint8_t GetFlags() const noexcept { return m_flags; }
@@ -189,7 +197,6 @@
         Assume(m_flags);
         return m_next;
     }
-
     //! Only call Prev when this entry is either DIRTY or FRESH
     inline CoinsCachePair* Prev() const noexcept {
         Assume(m_flags);
@@ -200,8 +207,7 @@
     inline void SelfRef(CoinsCachePair& self) noexcept
     {
         Assume(&self.second == this);
-        m_prev = &self;
-        m_next = &self;
+        SetPrevNext(&self, &self);
         // Set sentinel to DIRTY so we can call Next on it
         m_flags = DIRTY;
     }
```

(please find better names, I was uninspired)